File: ios/YamabikoChat/Features/Fusion/FusionTypes.swift
import Foundation

enum FusionPhase: String, Codable, Sendable {
    case panel
    case judge
    case synthesizer
}

enum FusionPanelChipState: String, Sendable, Equatable {
    case pending
    case running
    case succeeded
    case failed
}

struct FusionPanelChipStatus: Sendable, Equatable {
    var modelId: String
    var provider: String
    var state: FusionPanelChipState
    var toolActivity: ToolActivityPayload? = nil
}

struct FusionProgressSnapshot: Sendable, Equatable {
    var phase: FusionPhase
    var panels: [FusionPanelChipStatus]
    var completedPanelCount: Int
    var totalPanelCount: Int
    var substatus: String?

    static func panelPhase(
        panels: [FusionPanelChipStatus],
        substatus: String? = nil
    ) -> FusionProgressSnapshot {
        let completed = panels.filter {
            $0.state == .succeeded || $0.state == .failed
        }.count
        return FusionProgressSnapshot(
            phase: .panel,
            panels: panels,
            completedPanelCount: completed,
            totalPanelCount: panels.count,
            substatus: substatus
        )
    }

    static func initialPanels(from request: FusionRequest) -> [FusionPanelChipStatus] {
        request.panelModels.map { panel in
            FusionPanelChipStatus(
                modelId: panel.modelId,
                provider: panel.provider.uppercased(),
                state: .running
            )
        }
    }

    static func phaseOnly(
        _ phase: FusionPhase,
        panels: [FusionPanelChipStatus],
        substatus: String? = nil
    ) -> FusionProgressSnapshot {
        let completed = panels.filter {
            $0.state == .succeeded || $0.state == .failed
        }.count
        return FusionProgressSnapshot(
            phase: phase,
            panels: panels,
            completedPanelCount: completed,
            totalPanelCount: panels.count,
            substatus: substatus
        )
    }

    func applyingPanelResult(_ result: PanelResult) -> FusionProgressSnapshot {
        var updatedPanels = panels
        if let index = updatedPanels.firstIndex(where: { $0.modelId == result.modelId }) {
            updatedPanels[index].state = result.success ? .succeeded : .failed
            updatedPanels[index].toolActivity = result.toolActivity
        }
        return Self.panelPhase(panels: updatedPanels, substatus: substatus)
    }
}

enum FusionTaskType: String, Codable, Sendable, CaseIterable {
    case research
    case coding
    case auto
}

enum FusionConfidence: String, Codable, Sendable {
    case low
    case medium
    case high
}

struct PanelModelConfig: Codable, Sendable, Equatable {
    var modelId: String
    var provider: String
    var temperature: Double?
    /// Legacy preset field retained for decoding only; Fusion does not enforce it.
    var timeoutMs: Int?
    var role: String?
}

struct FusionRequest: Codable, Sendable, Equatable {
    var userPrompt: String
    var systemPrompt: String?
    var panelModels: [PanelModelConfig]
    var judgeModel: PanelModelConfig
    var synthesizerModel: PanelModelConfig
    var preset: String
    /// Legacy preset field retained for compatibility; Fusion does not enforce it.
    var timeoutMs: Int
    var allowWebSearch: Bool
    var taskType: FusionTaskType
    var metadata: [String: String]
}

struct PanelResult: Codable, Sendable, Equatable {
    var modelId: String
    var provider: String
    var success: Bool
    var content: String
    var error: String?
    var latencyMs: Int64
    var inputTokens: Int?
    var outputTokens: Int?
    var cost: Double?
    var toolCalls: [ToolCall]?
    var toolActivity: ToolActivityPayload? = nil
    var finishReason: String?
    var role: String?
}

struct JudgeContradiction: Codable, Sendable, Equatable {
    var topic: String
    var positions: [String]
    var likelyResolution: String
    var confidence: FusionConfidence
}

struct JudgeUniqueInsight: Codable, Sendable, Equatable {
    var model: String
    var insight: String
    var useInFinal: Bool
}

struct JudgeSuspectedError: Codable, Sendable, Equatable {
    var model: String
    var claim: String
    var reason: String
}

struct JudgeStrongestPart: Codable, Sendable, Equatable {
    var model: String
    var part: String
    var reason: String
}

struct JudgeAnalysis: Codable, Sendable, Equatable {
    var consensus: [String]
    var contradictions: [JudgeContradiction]
    var uniqueInsights: [JudgeUniqueInsight]
    var coverageGaps: [String]
    var suspectedErrors: [JudgeSuspectedError]
    var sourceQualityIssues: [String]
    var strongestAnswerParts: [JudgeStrongestPart]
    var recommendedFinalPosition: String
    var confidence: FusionConfidence
    var notes: String?
}

struct JudgePhaseResult: Codable, Sendable, Equatable {
    var analysis: JudgeAnalysis?
    var rawJSON: String?
    var parseSucceeded: Bool
    var latencyMs: Int64
    var inputTokens: Int?
    var outputTokens: Int?
    var cost: Double?
    var error: String?
}

struct SynthesisPhaseResult: Codable, Sendable, Equatable {
    var modelId: String
    var provider: String
    var success: Bool
    var content: String
    var latencyMs: Int64
    var inputTokens: Int?
    var outputTokens: Int?
    var cost: Double?
    var error: String?
}

struct FusionTrace: Codable, Sendable, Equatable {
    var requestId: String
    var preset: String
    var startedAtMs: Int64
    var completedAtMs: Int64?
    var panelResults: [PanelResult]
    var judgeResult: JudgePhaseResult?
    var synthesisResult: SynthesisPhaseResult?
    var totalLatencyMs: Int64?
    var totalCost: Double?
    var failedModels: [String]
    var status: String
    var userPrompt: String?
    var finalAnswer: String?
}

struct FusionContext: Sendable, Equatable {
    static let maxFusionDepth = 1

    var fusionDepth: Int
    var debugMode: Bool
    var logPrompts: Bool
    var conversationId: Int64?

    init(
        fusionDepth: Int = 0,
        debugMode: Bool = false,
        logPrompts: Bool = false,
        conversationId: Int64? = nil
    ) {
        self.fusionDepth = fusionDepth
        self.debugMode = debugMode
        self.logPrompts = logPrompts
        self.conversationId = conversationId
    }
}

struct FusionRunOptions: Sendable, Equatable {
    var taskType: FusionTaskType
    var systemPrompt: String?
    var debugMode: Bool
    var logPrompts: Bool
    var fusionDepth: Int
    var conversationId: Int64?
    var allowWebSearch: Bool?

    init(
        taskType: FusionTaskType = .auto,
        systemPrompt: String? = nil,
        debugMode: Bool = false,
        logPrompts: Bool = false,
        fusionDepth: Int = 0,
        conversationId: Int64? = nil,
        allowWebSearch: Bool? = nil
    ) {
        self.taskType = taskType
        self.systemPrompt = systemPrompt
        self.debugMode = debugMode
        self.logPrompts = logPrompts
        self.fusionDepth = fusionDepth
        self.conversationId = conversationId
        self.allowWebSearch = allowWebSearch
    }
}

struct FusionRunResult: Sendable, Equatable {
    var finalAnswer: String
    var traceId: String
    var judgeAnalysis: JudgeAnalysis?
    var rawPanelResults: [PanelResult]?
    var totalLatencyMs: Int64?
    var totalCost: Double?
}

struct FusionJudgeOutcome: Sendable {
    var trace: FusionTrace
    var synthesisRequest: ProviderRequest
    var synthesizerProvider: String
    var synthesizerModel: PanelModelConfig
    var panelTokenUsages: [(provider: String, model: String, usage: ProviderUsage?, requestType: String)]
    var judgeTokenUsage: (provider: String, model: String, usage: ProviderUsage?)?
}

enum FusionError: LocalizedError, Sendable {
    case allPanelsFailed(panelResults: [PanelResult])
    case presetNotFound(String)
    case invalidPreset(String)
    case recursionLimitExceeded
    case serviceDeallocated

    var errorDescription: String? {
        switch self {
        case .allPanelsFailed:
            return L10n.text("Fusion: すべてのパネルモデルが失敗しました。")
        case let .presetNotFound(name):
            return L10n.format("Fusion: プリセット '%@' が見つかりません。", name)
        case let .invalidPreset(reason):
            return L10n.format("Fusion: 無効なプリセット — %@", reason)
        case .recursionLimitExceeded:
            return L10n.text("Fusion: 再帰実行の上限に達しました。")
        case .serviceDeallocated:
            return L10n.text("Fusion: サービスが解放されました。")
        }
    }
}

// MARK: - Judge JSON DTO (snake_case from model output)

struct JudgeAnalysisDTO: Codable, Sendable {
    var consensus: [String]?
    var contradictions: [JudgeContradictionDTO]?
    var uniqueInsights: [JudgeUniqueInsightDTO]?
    var coverageGaps: [String]?
    var suspectedErrors: [JudgeSuspectedErrorDTO]?
    var sourceQualityIssues: [String]?
    var strongestAnswerParts: [JudgeStrongestPartDTO]?
    var recommendedFinalPosition: String?
    var confidence: String?
    var notes: String?
}

struct JudgeContradictionDTO: Codable, Sendable {
    var topic: String?
    var positions: [String]?
    var likelyResolution: String?
    var confidence: String?
}

struct JudgeUniqueInsightDTO: Codable, Sendable {
    var model: String?
    var insight: String?
    var useInFinal: Bool?
}

struct JudgeSuspectedErrorDTO: Codable, Sendable {
    var model: String?
    var claim: String?
    var reason: String?
}

struct JudgeStrongestPartDTO: Codable, Sendable {
    var model: String?
    var part: String?
    var reason: String?
}

extension JudgeAnalysisDTO {
    func toJudgeAnalysis() -> JudgeAnalysis {
        JudgeAnalysis(
            consensus: consensus ?? [],
            contradictions: (contradictions ?? []).map {
                JudgeContradiction(
                    topic: $0.topic ?? "",
                    positions: $0.positions ?? [],
                    likelyResolution: $0.likelyResolution ?? "",
                    confidence: FusionConfidence(rawValue: ($0.confidence ?? "medium").lowercased()) ?? .medium
                )
            },
            uniqueInsights: (uniqueInsights ?? []).map {
                JudgeUniqueInsight(
                    model: $0.model ?? "",
                    insight: $0.insight ?? "",
                    useInFinal: $0.useInFinal ?? true
                )
            },
            coverageGaps: coverageGaps ?? [],
            suspectedErrors: (suspectedErrors ?? []).map {
                JudgeSuspectedError(
                    model: $0.model ?? "",
                    claim: $0.claim ?? "",
                    reason: $0.reason ?? ""
                )
            },
            sourceQualityIssues: sourceQualityIssues ?? [],
            strongestAnswerParts: (strongestAnswerParts ?? []).map {
                JudgeStrongestPart(
                    model: $0.model ?? "",
                    part: $0.part ?? "",
                    reason: $0.reason ?? ""
                )
            },
            recommendedFinalPosition: recommendedFinalPosition ?? "",
            confidence: FusionConfidence(rawValue: (confidence ?? "medium").lowercased()) ?? .medium,
            notes: notes
        )
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/fusion/FusionTypes.kt
package com.porarri.yamabikochat.data.fusion

import com.porarri.yamabikochat.data.model.ProviderRequest
import com.porarri.yamabikochat.data.model.ProviderRequestMessage
import com.porarri.yamabikochat.data.model.ProviderUsage
import com.porarri.yamabikochat.data.model.ToolCall
import com.porarri.yamabikochat.data.local.ToolActivityPayload
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class FusionPhase {
    panel,
    judge,
    synthesizer,
    fallback
}

enum class FusionPanelChipState {
    pending,
    running,
    succeeded,
    failed
}

data class FusionPanelChipStatus(
    val modelId: String,
    val provider: String,
    val state: FusionPanelChipState,
    val toolActivity: ToolActivityPayload? = null
)

data class FusionProgressSnapshot(
    val phase: FusionPhase,
    val panels: List<FusionPanelChipStatus>,
    val completedPanelCount: Int,
    val totalPanelCount: Int,
    val substatus: String? = null
) {
    companion object {
        fun panelPhase(
            panels: List<FusionPanelChipStatus>,
            substatus: String? = null
        ): FusionProgressSnapshot {
            val completed = panels.count {
                it.state == FusionPanelChipState.succeeded || it.state == FusionPanelChipState.failed
            }
            return FusionProgressSnapshot(
                phase = FusionPhase.panel,
                panels = panels,
                completedPanelCount = completed,
                totalPanelCount = panels.size,
                substatus = substatus
            )
        }

        fun initialPanels(from: FusionRequest): List<FusionPanelChipStatus> =
            from.panelModels.map { panel ->
                FusionPanelChipStatus(
                    modelId = panel.modelId,
                    provider = panel.provider.uppercase(),
                    state = FusionPanelChipState.running
                )
            }

        fun phaseOnly(
            phase: FusionPhase,
            panels: List<FusionPanelChipStatus>,
            substatus: String? = null
        ): FusionProgressSnapshot {
            val completed = panels.count {
                it.state == FusionPanelChipState.succeeded || it.state == FusionPanelChipState.failed
            }
            return FusionProgressSnapshot(
                phase = phase,
                panels = panels,
                completedPanelCount = completed,
                totalPanelCount = panels.size,
                substatus = substatus
            )
        }
    }

    fun applyingPanelResult(result: PanelResult): FusionProgressSnapshot {
        val updatedPanels = panels.map { chip ->
            if (chip.modelId == result.modelId) {
                chip.copy(
                    state = if (result.success) {
                        FusionPanelChipState.succeeded
                    } else {
                        FusionPanelChipState.failed
                    },
                    toolActivity = result.toolActivity
                )
            } else {
                chip
            }
        }
        return panelPhase(panels = updatedPanels, substatus = substatus)
    }
}

@Serializable
enum class FusionTaskType {
    @SerialName("research")
    research,
    @SerialName("coding")
    coding,
    @SerialName("auto")
    auto
}

@Serializable
enum class FusionConfidence {
    high,
    medium,
    low;

    companion object {
        fun fromRaw(raw: String?): FusionConfidence = when (raw?.trim()?.lowercase()) {
            "high" -> high
            "medium" -> medium
            "low" -> low
            else -> medium
        }
    }
}

@Serializable
data class PanelModelConfig(
    val modelId: String,
    val provider: String,
    val temperature: Double? = null,
    val timeoutMs: Int? = null,
    val role: String? = "generalist"
)

@Serializable
data class FusionRequest(
    val userPrompt: String,
    val systemPrompt: String? = null,
    val panelModels: List<PanelModelConfig>,
    val judgeModel: PanelModelConfig,
    val synthesizerModel: PanelModelConfig,
    val fallbackModel: PanelModelConfig? = null,
    val preset: String = "custom",
    val maxPanelTokens: Int = 4096,
    val maxJudgeTokens: Int = 2048,
    val maxSynthesizerTokens: Int = 4096,
    val timeoutMs: Int = 120_000,
    val allowWebSearch: Boolean = true,
    val taskType: FusionTaskType = FusionTaskType.auto,
    val metadata: Map<String, String> = emptyMap()
)

@Serializable
data class SerializableToolCall(
    val id: String,
    val name: String,
    val argumentsJSON: String
)

fun ToolCall.toSerializable(): SerializableToolCall =
    SerializableToolCall(id = id, name = name, argumentsJSON = argumentsJSON)

@Serializable
data class PanelResult(
    val modelId: String,
    val provider: String,
    val success: Boolean,
    val content: String,
    val error: String? = null,
    val latencyMs: Long,
    val inputTokens: Int? = null,
    val outputTokens: Int? = null,
    val cost: Double? = null,
    val toolCalls: List<SerializableToolCall>? = null,
    val toolActivity: ToolActivityPayload? = null,
    val finishReason: String? = null,
    val role: String = "generalist"
)

@Serializable
data class JudgeContradiction(
    val topic: String,
    val positions: List<String>,
    val likelyResolution: String,
    val confidence: FusionConfidence
)

@Serializable
data class JudgeUniqueInsight(
    val model: String,
    val insight: String,
    val useInFinal: Boolean
)

@Serializable
data class JudgeSuspectedError(
    val model: String,
    val claim: String,
    val reason: String
)

@Serializable
data class JudgeStrongestPart(
    val model: String,
    val part: String,
    val reason: String
)

@Serializable
data class JudgeAnalysis(
    val consensus: List<String>,
    val contradictions: List<JudgeContradiction>,
    val uniqueInsights: List<JudgeUniqueInsight>,
    val coverageGaps: List<String>,
    val suspectedErrors: List<JudgeSuspectedError>,
    val sourceQualityIssues: List<String>,
    val strongestAnswerParts: List<JudgeStrongestPart>,
    val recommendedFinalPosition: String,
    val confidence: FusionConfidence,
    val notes: String? = null
)

@Serializable
data class JudgePhaseResult(
    val analysis: JudgeAnalysis? = null,
    val rawJSON: String? = null,
    val parseSucceeded: Boolean,
    val latencyMs: Long,
    val inputTokens: Int? = null,
    val outputTokens: Int? = null,
    val cost: Double? = null,
    val error: String? = null
)

@Serializable
data class SynthesisPhaseResult(
    val modelId: String,
    val provider: String,
    val success: Boolean,
    val content: String,
    val latencyMs: Long,
    val inputTokens: Int? = null,
    val outputTokens: Int? = null,
    val cost: Double? = null,
    val error: String? = null,
    val usedFallback: Boolean
)

@Serializable
data class FusionTrace(
    val requestId: String,
    val preset: String,
    val startedAtMs: Long,
    val completedAtMs: Long? = null,
    val panelResults: List<PanelResult>,
    val judgeResult: JudgePhaseResult? = null,
    val synthesisResult: SynthesisPhaseResult? = null,
    val totalLatencyMs: Long? = null,
    val totalCost: Double? = null,
    val failedModels: List<String>,
    val status: String,
    val userPrompt: String? = null,
    val finalAnswer: String? = null
)

data class FusionContext(
    val fusionDepth: Int = 0,
    val debugMode: Boolean = false,
    val logPrompts: Boolean = false,
    val conversationId: Long? = null,
    val clientToolsAllowed: Boolean = true
) {
    companion object {
        const val MAX_FUSION_DEPTH = 1
    }
}

data class FusionRunOptions(
    val taskType: FusionTaskType = FusionTaskType.auto,
    val systemPrompt: String? = null,
    val debugMode: Boolean = false,
    val logPrompts: Boolean = false,
    val fusionDepth: Int = 0,
    val conversationId: Long? = null,
    val allowWebSearch: Boolean? = null
)

data class FusionRunResult(
    val finalAnswer: String,
    val traceId: String,
    val judgeAnalysis: JudgeAnalysis? = null,
    val rawPanelResults: List<PanelResult>? = null,
    val totalLatencyMs: Long? = null,
    val totalCost: Double? = null
)

data class FusionTokenUsageRecord(
    val provider: String,
    val model: String,
    val usage: ProviderUsage?,
    val requestType: String
)

data class FusionJudgeOutcome(
    val trace: FusionTrace,
    val synthesisRequest: ProviderRequest,
    val synthesizerProvider: String,
    val synthesizerModel: PanelModelConfig,
    val staticFallbackAnswer: String,
    val panelTokenUsages: List<FusionTokenUsageRecord>,
    val judgeTokenUsage: FusionTokenUsageRecord?
)

sealed class FusionError(message: String) : Exception(message) {
    data class AllPanelsFailed(val panelResults: List<PanelResult>) :
        FusionError("Fusion: すべてのパネルモデルが失敗しました。")

    data class PresetNotFound(val name: String) :
        FusionError("Fusion: プリセット '$name' が見つかりません。")

    data class InvalidPreset(val reason: String) :
        FusionError("Fusion: 無効なプリセット — $reason")

    data object ServiceDeallocated :
        FusionError("Fusion: サービスが解放されました。")
}

@Serializable
data class JudgeAnalysisDto(
    val consensus: List<String>? = null,
    val contradictions: List<JudgeContradictionDto>? = null,
    @SerialName("unique_insights")
    val uniqueInsights: List<JudgeUniqueInsightDto>? = null,
    @SerialName("coverage_gaps")
    val coverageGaps: List<String>? = null,
    @SerialName("suspected_errors")
    val suspectedErrors: List<JudgeSuspectedErrorDto>? = null,
    @SerialName("source_quality_issues")
    val sourceQualityIssues: List<String>? = null,
    @SerialName("strongest_answer_parts")
    val strongestAnswerParts: List<JudgeStrongestPartDto>? = null,
    @SerialName("recommended_final_position")
    val recommendedFinalPosition: String? = null,
    val confidence: String? = null,
    val notes: String? = null
)

@Serializable
data class JudgeContradictionDto(
    val topic: String? = null,
    val positions: List<String>? = null,
    @SerialName("likely_resolution")
    val likelyResolution: String? = null,
    val confidence: String? = null
)

@Serializable
data class JudgeUniqueInsightDto(
    val model: String? = null,
    val insight: String? = null,
    @SerialName("use_in_final")
    val useInFinal: Boolean? = null
)

@Serializable
data class JudgeSuspectedErrorDto(
    val model: String? = null,
    val claim: String? = null,
    val reason: String? = null
)

@Serializable
data class JudgeStrongestPartDto(
    val model: String? = null,
    val part: String? = null,
    val reason: String? = null
)

fun JudgeAnalysisDto.toJudgeAnalysis(): JudgeAnalysis =
    JudgeAnalysis(
        consensus = consensus.orEmpty(),
        contradictions = contradictions.orEmpty().map {
            JudgeContradiction(
                topic = it.topic.orEmpty(),
                positions = it.positions.orEmpty(),
                likelyResolution = it.likelyResolution.orEmpty(),
                confidence = FusionConfidence.fromRaw(it.confidence)
            )
        },
        uniqueInsights = uniqueInsights.orEmpty().map {
            JudgeUniqueInsight(
                model = it.model.orEmpty(),
                insight = it.insight.orEmpty(),
                useInFinal = it.useInFinal ?: true
            )
        },
        coverageGaps = coverageGaps.orEmpty(),
        suspectedErrors = suspectedErrors.orEmpty().map {
            JudgeSuspectedError(
                model = it.model.orEmpty(),
                claim = it.claim.orEmpty(),
                reason = it.reason.orEmpty()
            )
        },
        sourceQualityIssues = sourceQualityIssues.orEmpty(),
        strongestAnswerParts = strongestAnswerParts.orEmpty().map {
            JudgeStrongestPart(
                model = it.model.orEmpty(),
                part = it.part.orEmpty(),
                reason = it.reason.orEmpty()
            )
        },
        recommendedFinalPosition = recommendedFinalPosition.orEmpty(),
        confidence = FusionConfidence.fromRaw(confidence),
        notes = notes
    )

/** History content for panel phase (role + text + optional attachment paths). */
data class FusionHistoryMessage(
    val role: String,
    val text: String,
    val attachments: List<String> = emptyList()
)
